File: go/300.500/433.MinimumGeneticMutation.go
package main

import "fmt"

var (
	layer    int
	minlayer int
	mend     *string
	used     []bool
	mbank    []string
)

func minMutation(start string, end string, bank []string) int {
	layer = 0
	minlayer = len(bank) + 1
	mend = &end
	used = make([]bool, len(bank))
	mbank = bank
	mutation(&start)
	if minlayer > len(bank) {
		return -1
	}
	return minlayer
}

func mutation(start *string) {
	if layer >= minlayer {
		return
	}
	layer++
	for i, x := range mbank {
		if used[i] {
			continue
		}
		if different(start, &x) {
			if *mend == x {
				minlayer = layer
				layer--
				return
			}
			used[i] = true
			mutation(&x)
			used[i] = false
		}
	}
	layer--
}

func different(left, right *string) bool {
	cnt := 0
	for i := 0; i < 8; i++ {
		if (*left)[i] != (*right)[i] {
			cnt++
		}
	}
	return cnt == 1
}

func main() {
	fmt.Println(minMutation("AACCGGTT", "AACCGGTA", []string{"AACCGGTA"}), 1)
	fmt.Println(minMutation("AAAACCCC", "CCCCCCCC", []string{"AAAACCCA", "AAACCCCA", "AACCCCCA", "AACCCCCC", "ACCCCCCC", "CCCCCCCC", "AAACCCCC", "AACCCCCC"}), 4)
	fmt.Println(minMutation("AACCGGTT", "AAACGGTA", []string{"AACCGGTA", "AACCGCTA", "AAACGGTA"}), 2)
	fmt.Println(minMutation("AACCGGTT", "AAACGGTA", []string{"AACCGATT", "AACCGATA", "AAACGATA", "AAACGGTA"}), 4)
	fmt.Println(minMutation("AACCTTGG", "AATTCCGG", []string{"AATTCCGG", "AACCTGGG", "AACCCCGG", "AACCTACC"}), -1)
	fmt.Println(minMutation("AAAAACCC", "AACCCCCC", []string{"AAAACCCC", "AAACCCCC", "AACCCCCC"}), 3)
}

File: 300.500/433.MinimumGeneticMutation.cpp
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <vector>
#include <array>
using namespace std;

class MinimumGeneticMutation {
public:
    int minMutation(string start, string end, vector<string>& bank) {
        vector<bool> used(bank.size(), false);
        layer = 0;
        mend = end;
        minlayer = bank.size() + 1;
        mutation(start, bank, used);
        if (minlayer > bank.size())
            return -1;
        return minlayer;
    }
private:
    void mutation(string& start, vector<string>& bank, vector<bool>& used) {
        if (layer >= minlayer) {
            return;
        }
        layer += 1;
        for (int i = 0;i < bank.size();++i) {
            if (used[i])
                continue;   //已经使用过了
            if (oneDifferent(start, bank[i])) {
                //cout << " cur choose: " << i << " layer: " << layer;
                if (mend.compare(bank[i]) == 0) {
                    //cout << " end layer:" << layer << endl;
                    minlayer = layer;
                    layer -= 1;
                    return;
                }
                used[i] = true;
                mutation(bank[i], bank, used);
                used[i] = false;
            }
        }
        layer -= 1;
    }
    bool oneDifferent(string& left, string& right) {
        int cnt = 0;
        for (int i = 0; i < 8;++i) {
            if (left[i] != right[i])
                ++cnt;
        }
        return cnt == 1;
    }
private:
    int layer{};
    int minlayer{};
    string mend{};
};

int main(int argc, char const* argv[]) {
    MinimumGeneticMutation cls;
    cout << cls.minMutation("AAAACCCC", "CCCCCCCC", vector<string>{"AAAACCCA","AAACCCCA","AACCCCCA","AACCCCCC","ACCCCCCC","CCCCCCCC","AAACCCCC","AACCCCCC"}) << " -> 4" << endl;
    cout << cls.minMutation("AACCGGTT", "AAACGGTA", vector<string>{"AACCGGTA", "AACCGCTA", "AAACGGTA"}) << " -> 2" << endl;
    cout << cls.minMutation("AACCGGTT", "AAACGGTA", vector<string>{"AACCGATT", "AACCGATA", "AAACGATA", "AAACGGTA"}) << " -> 4" << endl;
    cout << cls.minMutation("AACCGGTT", "AACCGGTA", vector<string>{"AACCGGTA"}) << " -> 1" << endl;
    cout << cls.minMutation("AACCTTGG", "AATTCCGG", vector<string>{"AATTCCGG", "AACCTGGG", "AACCCCGG", "AACCTACC"}) << " -> -1" << endl;
    cout << cls.minMutation("AAAAACCC", "AACCCCCC", vector<string>{"AAAACCCC", "AAACCCCC", "AACCCCCC"}) << " -> 3" << endl;
    return 0;
}
